walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,104 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.HashSet;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
-}
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+            return Arrays.asList();
+        }
+        if (events.isEmpty()) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+
+        int differentAttendees = 1;
+
+        List<TimeRange> available = Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TimeRange.END_OF_DAY, true));
+        // Arrays.asList(TimeRange.fromStartEnd(TimeRange.START_OF_DAY, TIME_0800AM, false),
+        //     TimeRange.fromStartEnd(TIME_0830AM, TIME_0900AM, false),
+        //     TimeRange.fromStartEnd(TIME_0930AM, TimeRange.END_OF_DAY, true));
+        for (Event event : events) {
+            if (attendeesOverlap(event.getAttendees(), request.getAttendees())) {
+                available = remove(available, event.getWhen());
+                differentAttendees = 0;
+            }
+        }
+
+        if (differentAttendees == 1) {
+            return Arrays.asList(TimeRange.WHOLE_DAY);
+        }
+        return available;
+    }
+
+    public List<TimeRange> remove(Collection<TimeRange> collectionRanges, TimeRange rangeToRemove) {
+        int index = -1;
+        ArrayList<TimeRange> ranges = new ArrayList<>(collectionRanges);
+        for (TimeRange range : ranges) {
+            
+            index++;
```

Comment:
I'm suspicious of the `index` manipulation that's happening here.

I *think* your code is correct, but this type of careful manipulation of lists by index is very hard to verify when reading the code.

You are also removing and inserting items from the array that you are looping over -- this typically results in a `ConcurrentModificationException` when the iterator notices that the underlying array has changed.

Looks like you narrowly avoid this case by adding a `break;` after changing the collection size. This may actually be a bug if there are other items to consider after a "split" of a TimeRange.

I would suggest that you think about working with *two* lists, one input and one output, and never modify the list that's being iterated over.

I think you can prove that the `ranges` are *never* overlapping (since they start out containing just `WHOLE_DAY` and are only ever split/shortened into smaller ranges).
If that's true, then you can process:

```
List<..> result =...;
for (r in ranges) {
 if (case1) {
 // remove it (don't add it to result)
 } else if (case2 || case 3) {
 // shorten it in the front or back, add it to result
 result.add(shorter)
 } else if (case4) {
 // split it, add both new segments to result
 result.add(front);
 result.add(back);
 }
}
```